File: nimon-backend/src/common/validation/text-normalization.ts
/** Trim leading/trailing whitespace; treat null/undefined as empty string. */
export function trimText(value: string | null | undefined): string {
  if (value == null) return '';
  return String(value).trim();
}

/** Collapse internal runs of whitespace to a single space (does not trim ends). */
export function collapseSpaces(value: string): string {
  return value.replace(/\s+/g, ' ');
}

/** Trim, collapse spaces, remove line breaks as spaces for single-line fields. */
export function normalizeSingleLineText(value: string): string {
  const t = trimText(value);
  const noBreaks = t.replace(/[\r\n]+/g, ' ');
  return collapseSpaces(noBreaks).trim();
}

const SCRIPTISH =
  /<\/|\/?script|javascript:|on\w+\s*=|<iframe|<object|<embed|<svg[\s\S]*on/i;

export function containsHtmlOrScript(value: string): boolean {
  if (!value) return false;
  if (SCRIPTISH.test(value)) return true;
  if (/<[a-z][\s\S]*>/i.test(value)) return true;
  return false;
}

const URL_PATTERN =
  '(?:https?:\\/\\/|www\\.)[^\\s]+|[a-z0-9][a-z0-9-]*\\.[a-z]{2,}\\b[^\\s]*';

export function containsUrl(value: string): boolean {
  if (!value) return false;
  return new RegExp(URL_PATTERN, 'gi').test(value);
}

/** Rough emoji count (extended pictographic + keycap sequences). */
export function countEmojis(value: string): number {
  if (!value) return 0;
  const re = /\p{Extended_Pictographic}/gu;
  const m = value.match(re);
  return m?.length ?? 0;
}

export function isOnlyNumbers(value: string): boolean {
  const t = trimText(value);
  if (!t) return false;
  return /^[0-9０-９]+$/.test(t);
}

/** Letters/numbers only → not "only symbols"; this flags symbol-only noise. */
export function isOnlySymbols(value: string): boolean {
  const t = trimText(value);
  if (!t) return false;
  if (/[\p{L}\p{N}]/u.test(t)) return false;
  return /^[\p{S}\p{P}\s]+$/u.test(t);
}

export function hasExcessiveRepeatedCharacters(
  value: string,
  maxRepeat = 4,
): boolean {
  if (!value || maxRepeat < 2) return false;
  const re = new RegExp(`(.)\\1{${maxRepeat},}`, 'u');
  return re.test(value);
}

export function countLines(value: string): number {
  if (!trimText(value)) return 0;
  const parts = value.split(/\r\n|\r|\n/);
  return parts.length;
}

/** Unicode code-point length (better than UTF-16 .length for many scripts). */
export function charLength(value: string): number {
  return [...value].length;
}

export function countUrls(value: string): number {
  if (!value) return 0;
  const m = value.match(new RegExp(URL_PATTERN, 'gi'));
  return m?.length ?? 0;
}

export function countHashtags(value: string): number {
  if (!value) return 0;
  const m = value.match(/#[\p{L}\p{N}_]+/gu);
  return m?.length ?? 0;
}

File: lib/core/validation/text_normalization.dart
import 'package:characters/characters.dart';

String trimText(String? value) => value == null ? '' : value.trim();

String collapseSpaces(String value) => value.replaceAll(RegExp(r'\s+'), ' ');

String normalizeSingleLineText(String value) {
  final t = trimText(value);
  final noBreaks = t.replaceAll(RegExp(r'[\r\n]+'), ' ');
  return collapseSpaces(noBreaks).trim();
}

final _scriptish = RegExp(
  r'<\/|\/?script|javascript:|on\w+\s*=|<iframe|<object|<embed|<svg[\s\S]*on',
  caseSensitive: false,
);

bool containsHtmlOrScript(String value) {
  if (value.isEmpty) return false;
  if (_scriptish.hasMatch(value)) return true;
  if (RegExp(r'<[a-z][\s\S]*>', caseSensitive: false).hasMatch(value)) {
    return true;
  }
  return false;
}

const _urlPattern =
    r'(?:https?:\/\/|www\.)[^\s]+|[a-z0-9][a-z0-9-]*\.[a-z]{2,}\b[^\s]*';

bool containsUrl(String value) {
  if (value.isEmpty) return false;
  return RegExp(_urlPattern, caseSensitive: false).hasMatch(value);
}

int countUrls(String value) {
  if (value.isEmpty) return 0;
  final m = RegExp(_urlPattern, caseSensitive: false).allMatches(value);
  return m.length;
}

/// Uses unicode extended pictographic class when supported by the VM.
final _emojiRe = RegExp(r'\p{Extended_Pictographic}', unicode: true);

int countEmojis(String value) {
  if (value.isEmpty) return 0;
  return _emojiRe.allMatches(value).length;
}

bool isOnlyNumbers(String value) {
  final t = trimText(value);
  if (t.isEmpty) return false;
  return RegExp(r'^[0-9０-９]+$').hasMatch(t);
}

bool isOnlySymbols(String value) {
  final t = trimText(value);
  if (t.isEmpty) return false;
  if (RegExp(r'[\p{L}\p{N}]', unicode: true).hasMatch(t)) return false;
  return RegExp(r'^[\p{S}\p{P}\s]+$', unicode: true).hasMatch(t);
}

bool hasExcessiveRepeatedCharacters(String value, [int maxRepeat = 4]) {
  if (value.isEmpty || maxRepeat < 2) return false;
  final re = RegExp('(.)\\1{$maxRepeat,}', unicode: true);
  return re.hasMatch(value);
}

int countLines(String value) {
  if (trimText(value).isEmpty) return 0;
  return value.split(RegExp(r'\r\n|\r|\n')).length;
}

int charLength(String value) => value.characters.length;

int countHashtags(String value) {
  if (value.isEmpty) return 0;
  final m = RegExp(r'#[\p{L}\p{N}_]+', unicode: true).allMatches(value);
  return m.length;
}
